### PR TITLE
fix require presence for multiple challenges

### LINKED ISSUE #4122: Push require presence not working when another token is triggered aswell
### Top-level intent
Use a Push token with require presence and a hotp token in challenge-response mode.

What did you try to achieve?
Have the require presence feature working

### Steps to reproduce

1. Enroll tokens as stated above
2. Trigger them
3. Look at the challenge message/app to see that require presence was not triggered

### Expected outcome

Have the require presence feature working

### Actual outcome

Push did work just without require presence

### PR DESCRIPTION
**Pull request by nilsbehlen**

### CODE REVIEW

**Comment by nilsbehlen** [comment]
sorry lots of formatting

**Review by plettich — CHANGES_REQUESTED** [review]
Just the `nonce` issue... if it isn't required anymore, we should explain that.

File: privacyidea/lib/tokens/pushtoken.py [hunk 1]
```diff
@@ -684,23 +686,22 @@ def _api_endpoint_post(cls, request_data):
...
 
-            if challengeobject_list:
+            if challenges:
                 # There are valid challenges, so we check this signature
-                for chal in challengeobject_list:
-                    # verify the signature of the nonce
-                    sign_data = f"{challenge}|{serial}"
+                for challenge in challenges:
+                    # Re-construct the signature data and then verify the signature
+                    sign_data = f"{challenge.challenge}|{serial}"
```
[the remark this answers is not in the document]
Comment (nilsbehlen) [reply privacyidea/lib/tokens/pushtoken.py:704]: the nonce is pointless as a parameter in the request. the server should not use parameters given by a client to verify something (other than the signature, and in this case). the server uses the parameters from its database to check the signature and if the signature does not match, we raise an error. i do not think we should to tell the client that the nonce it used was wrong by comparing the one from the db with the one from the request.

**Comment by nilsbehlen** [comment]
i have a note hanging on my screen to update the doc with the changes/removal of the nonce parameter, i only need to do it

**Review by frankmer — APPROVED** [review]

**Comment by nilsbehlen** [comment]
i have only found the doc inside the pushtoken class. since the push response is handled over the generic `/ttype/` endpoint, there is no doc for that endpoint. i also adjusted the concept in the wiki.

File: privacyidea/lib/tokens/pushtoken.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py:704]: So basically we send the challenge data as the nonce :-) It is just not described like that everywhere.

File: privacyidea/lib/tokens/pushtoken.py [hunk 2]
```diff
@@ -879,7 +878,6 @@ def api_endpoint(cls, request, g):
               Host: https://yourprivacyideaserver
 
               serial=<token serial>
-              nonce=<the actual challenge>
               decline=1
               signature=<signature over {nonce}|{serial}|decline
```
Comment (plettich) [inline privacyidea/lib/tokens/pushtoken.py:882]: like above

**Review by plettich — CHANGES_REQUESTED** [review]

File: privacyidea/lib/tokens/pushtoken.py [hunk 3]
```diff
@@ -867,7 +867,6 @@ def api_endpoint(cls, request, g):
               Host: https://yourprivacyideaserver
 
               serial=<token serial>
-              nonce=<the actual challenge>
               signature=<signature over {nonce}|{serial}>
```
[the remark this answers is not in the document]
Comment (nilsbehlen) [reply privacyidea/lib/tokens/pushtoken.py:870]: like i said, we do not need that parameter any more, i so have removed it

**Pull request merged**